finesse/hardware/plugins/spectrometer/ftsw500_interface.py

```diff
@@ -0,0 +1,194 @@
+                logging.info(f"{msg.split('&')[1][:-1]}")
+                return None
+        else:
+            return None
+    else:
+        raise FTSW500Error("Unrecognised response")
+
+    if status == -1:
+        return SpectrometerStatus(1)
+    elif status in (0, 1, 2, 3):
+        return SpectrometerStatus(status)
+    else:
+        raise FTSW500Error("Unable to parse response")
+
+
+class FTSW500Interface(FTSW500InterfaceBase, description="FTSW500 spectrometer"):
+    """Interface for communicating with the FTSW500 program."""
+
+    def __init__(self) -> None:
+        """Create a new FTSW500Interface."""
+        super().__init__()
+
+        self._requester = socket(AF_INET, SOCK_STREAM)
+        self._requester.settimeout(FTSW500_TIMEOUT)
+
+        self._requester.connect((FTSW500_HOST, FTSW500_PORT))
+
+        self._status = SpectrometerStatus.UNDEFINED
+
+        self._status_timer = QTimer()
+        self._status_timer.timeout.connect(self._request_status)
+        self._status_timer.setInterval(int(FTSW500_POLLING_INTERVAL * 1000))
+        self._status_timer.setSingleShot(True)
+
+        self._request_status()
+
+    def _check_is_modal_dialog_open(self) -> bool:
+        """Query whether FTSW500 has a modal dialog open."""
+        self._requester.sendall(b"isModalMessageDisplayed\n")
+        data = self._requester.recv(1024)
+        if data != b"":
+            if data.decode().split("&")[1] == "true\n":
+                return True
+            else:
+                return False
+        else:
+            return False
+
+    def _check_is_nonmodal_dialog_open(self) -> bool:
+        """Query whether FTSW500 has a non-modal dialog open."""
+        self._requester.sendall(b"isNonModalMessageDisplayed\n")
+        data = self._requester.recv(1024)
+        if data != b"":
+            if data.decode().split("&")[1] == "true\n":
+                return True
+            else:
+                return False
+        else:
+            return False
+
+    def _regurgitate_modal_dialog_message(self) -> None:
+        """Obtain the last modal message displayed on FTSW500 and log it."""
+        if self._check_is_modal_dialog_open():
+            self._requester.sendall(b"getLastModalMessageDisplayed\n")
+            data = self._requester.recv(1024)
+            logging.info(f"FTSW500: {data.decode().split('&')[1][:-1]}")
+            self._requester.sendall(b"closeModalDialogMessage\n")
+            self._requester.recv(1024)
+
+    def _regurgitate_nonmodal_dialog_message(self) -> None:
+        """Obtain the last non-modal message displayed on FTSW500 and log it."""
+        if self._check_is_nonmodal_dialog_open():
+            self._requester.sendall(b"getLastNonModalMessageDisplayed\n")
+            data = self._requester.recv(1024)
+            logging.info(f"FTSW500: {data.decode().split('&')[1][:-1]}")
+            self._requester.sendall(b"closeNonModalDialogMessage\n")
+            self._requester.recv(1024)
```

Comment:
While the messages sent are different, the logic in both cases seems to be the same. I'd suggest you combine both and use wrappers for each of them. Something like:

```suggestion
    def _is_dialog_open(self, check_modal: bool) -> bool:
        """Query whether FTSW500 has a dialog open."""
        msg = "" if check_modal else "Non"
        self._requester.sendall(f"is{msg}ModalMessageDisplayed\n".encode("utf-8"))
        data = self._requester.recv(1024)
        if data != b"" and data.decode().split("&")[1] == "true\n":
            return True
        return False
        
    def _regurgitate_dialog_message(self, is_modal: bool) -> None:
        """Obtain the last message displayed on FTSW500 and log it."""
        msg = "" if is_modal else "Non"
        self._requester.sendall(f"getLast{msg}ModalMessageDisplayed\n".encode("utf-8"))
        data = self._requester.recv(1024)
        logging.info(f"FTSW500: {data.decode().split('&')[1][:-1]}")
        self._requester.sendall(f"close{msg}ModalDialogMessage\n".encode("utf-8"))
        self._requester.recv(1024)

    def _regurgitate_modal_dialog_message(self) -> None:
        """Obtain the last modal message displayed on FTSW500 and log it."""
        if self._is_dialog_open(check_modal=True)
            self._regurgitate_dialog_message(is_modal=True)

    def _regurgitate_non_modal_dialog_message(self) -> None:
        """Obtain the last non modal message displayed on FTSW500 and log it."""
        if self._is_dialog_open(check_modal=False)
            self._regurgitate_dialog_message(is_modal=False)
```

Reply:
I'm tempted to remove the handling of modal dialog messages from our interface, as these generally mean user input is required in FTSW500 directly.